scipy/stats/stats.py

```diff
@@ -8121,6 +8121,281 @@ def brunnermunzel(x, y, alternative="two-sided", distribution="t",
+    later proven to be applicable to any set of i.i.d. samples [2]_. The
+    computation is based on the observation that the probability of a quantile
+    :math:`q` to be larger than any sample :math:`x_m (1\leq m \leq N)` can be
+    computed as
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = 1 - \sum_{k=0}^{m-1} \binom{N}{k}
+        q^k(1-q)^{N-k}
+
+    Furthermore, these probabilities are symmetric, which allows to compute both
+    upper and lower bounds from the same computation:
+
+    .. math::
+
+        \mathbb{P}(x_m \leq q) = \mathbb{P}(x_{N-m+1} \geq 1-q).
+
+    The function computes confidence intervals for a given quantile and
+    confidence level based on `x`, which is either a set of samples
+    (one-dimensional array_like) or the number of samples available. The
+    confidence intervals are valid if and only if the samples are i.i.d.
+
+    Both one-sided and two-sided confidence intervals can be obtained (default
+    is one-sided). The function returns two values: either the bounds for the
+    two one-sided confidence intervals, or the lower and upper bounds of a
+    two-sided confidence interval.
+
+    If `x` is the set of samples, the return values are the sample values
+    corresponding with the bounds of the confidence interval. If `x` is an
+    integer, the function considers `x` as the sample size and returns the
+    indexes of the bounds for the confidence interval for this sample size. Note
+    that this is possible because the numerical values of the data are
+    irrelevant for this computing confidence interval using this method.
+
+    A similar function is available in the QuantileNPCI R package [3]_. The
+    foundation is the same, but it computes the confidence interval bounds by
+    doing interpolations between the sample values, whereas this function uses
+    only sample values as bounds. Thus, `confint_quantile` returns slightly more
+    conservative intervals (i.e., larger).
+
+    Parameters
+    ----------
+    x : array_like or int Array of samples, should be one-dimensional. If
+        integer, taken as the number of samples available (strictly positive)
+    quantile : float The quantile for which we want to compute the confidence 
+        interval. Must be strictly between 0 and 1. 
+    confidence : float The desired confidence level of the confidence interval. 
+        Must be strictly between 0 and 1.
+    type : {'one-sided', 'two-sided'}, optional
+        Defines the type of confidence interval computed. Default is
+        'one-sided'.
+
+          * 'one-sided' :
+
+          computes the best possible one-sided confidence intervals
+          (both lower and upper bounds) for the given quantile.
+
+          * 'two-sided' :
+
+          computes a two-sided confidence interval by combination of
+          two one-sided intervals. E.g., a 90% two-sided interval is
+          computed by combining two 95% one-sided intervals.
+
+    Returns
+    -------
```

Comment:
I need to think more about this. In hindsight, having stats functions return individual variables (or tuples) has made them difficult to extend in the future. I think we're going to want to return an object here. Is there anything else that object should contain (e.g. estimate of the quantile)? (This gets me thinking about whether this function should be actually be called `quantile` and basically be an extension of `np.quantile` that returns not only an estimate of the underlying distribution's quantile but also a confidence interval.)

Reply:
I don't have anything against returning an object (I quite like the idea actually). If we go down this way, then I would return an object called e.g. CI with pretty much all information. That is, with the following properties:
```
CI.quantile       # the estimated quantile
CI.conf_level     # the (desired) confidence level
CI.type           # one- or two-sided interval
CI.sample_size    # the size of the sample used to compute the CI
CI.lb             # the index of the CI lower bound
CI.ub             # the index of the CI upper bound
CI.lb_value       # the value of the CI lower bound (if available)
CI.ub_value       # the value of the CI upper bound (if available)
``` 

Regarding your suggestion to make this function an extension of `np.quantile`:
I think it makes sense, but I fear it would make the function usage somehow blurry/hard to find. My thinking is, many people do not use CIs because they don't know much about them and it's not (always) easy to find an implementation to compute them. In order to fix that, I think having a function that is explicitly made for computing CIs is better than mixing everything into a global `quantile` function that would do everything (multi-dim, different interpolations for empirical quantile, etc.)

Maybe more fundamentally, `np.quantile` deals with empirical quantiles, while here we aim to estimate distribution quantiles; now that I think about it this way, I become even more against merging the two. These two concepts are already confused enough :-)

Reply:
OK. Maybe `scipy.stats.quantile` is not what we want, but I think we'll need to think about the interface a little more. We don't really have an established pattern for confidence intervals to follow, so we want to structure things in a way that can be extended.

Reply:
I'm all in for that, and I'll gladly do the extra effort to make things more extendable. I'm just not most experience to judge/have an opinion on how that would look like.